packages/studio/src/components/ComponentTree.tsx

```diff
@@ -167,32 +204,118 @@ function useTree(): NodeModel<ComponentState>[] | undefined {
 }
 
 function useDropHandler() {
-  const updateComponentTree = useStudioStore(
-    (store) => store.actions.updateComponentTree
-  );
+  const [selectedComponentUUIDs, componentTree, updateComponentTree] =
+    useStudioStore((store) => {
+      return [
+        store.pages.selectedComponentUUIDs,
+        store.actions.getComponentTree(),
+        store.actions.updateComponentTree,
+      ];
+    });
 
-  const handleDrop = useCallback(
-    (tree: NodeModel<ComponentState>[]) => {
-      const updatedComponentTree = tree.map((n) => {
-        if (!n.data) {
-          throw new Error(
-            "Unable to handle drag and drop event in ComponentTree: " +
-              "No data passed into NodeModel<ComponentState> for node " +
-              JSON.stringify(n, null, 2)
-          );
-        }
-        const componentState: ComponentState = {
-          ...n.data,
-          parentUUID: n.parent.toString(),
-        };
-        if (componentState.parentUUID === ROOT_ID) {
-          delete componentState.parentUUID;
+  const handleSelectedDrop = useCallback(
+    (
+      updatedComponentTree: ComponentState[],
+      dragSourceId: string,
+      destinationParentId: string
+    ) => {
+      if (!componentTree) {
+        throw new Error(
+          "Unable to handle drag and drop event: component tree is undefined."
+        );
+      }
+      const hasNoChanges = isEqual(
+        ComponentTreeHelpers.mapComponentTreeParentsFirst(
+          updatedComponentTree,
+          (c) => c
+        ),
+        ComponentTreeHelpers.mapComponentTreeParentsFirst(
+          componentTree,
+          (c) => c
+        )
+      );
+      if (hasNoChanges) {
+        return;
+      }
+      const selectedUUIDsArray = Array.from(selectedComponentUUIDs);
+      const lowestParentUUID = ComponentTreeHelpers.getLowestParentUUID(
+        selectedUUIDsArray[0],
+        selectedUUIDsArray[selectedUUIDsArray.length - 1],
+        componentTree
+      );
+      const selectedComponents = componentTree
+        .filter((c) => selectedComponentUUIDs.has(c.uuid))
+        .map((c) => {
+          if (c.parentUUID !== lowestParentUUID) return c;
+          return updateComponentParentUUID(c, destinationParentId);
+        });
+      updatedComponentTree = updatedComponentTree.filter(
+        (c) => c.uuid === dragSourceId || !selectedComponentUUIDs.has(c.uuid)
+      );
+      let newDestinationIndex;
+      updatedComponentTree.forEach((c, index) => {
+        if (c.uuid === dragSourceId) {
+          newDestinationIndex = index;
         }
```

Comment:
can we use `Array.findIndex` to calculate `newDestinationIndex`?

Reply:
Done

Reply:
I think this still needs to be updated?